src/app/modules/service-providers/services/navigation/navigation.service.ts

```diff
@@ -1,16 +1,101 @@
+        routeDetails.routeLegs[newNextWpIndex].index <= currentPositionOnRoute.properties.index) {
+          newNextWpIndex++;
+        }
+        if (newNextWpIndex > nextWpIndex) {
+          this.store.dispatch(new SetNextWaypointIndexAction({ nextWaypointIndex: newNextWpIndex }));
+        }
 
+        const lineToNextWp = lineSlice(point([currentPositionOnRoute.geometry.coordinates[0],
+            currentPositionOnRoute.geometry.coordinates[1]
+          ]),
+          point(routeDetails.routeLegs[newNextWpIndex].coords),
+          lineString(routePoints));
+        const distanceToNextWp = length(lineToNextWp) * 1000;
+        this.store.dispatch(new SetNextWaypointDistanceAction({ nextWaypointDistance: distanceToNextWp }));
+      }),
+      withLatestFrom(this.store.pipe(select(getNavState))),
+      filter(([[[currentPosition, routePoints], nextWpIndex, routeDetails], stateData]) => stateData.previousPosition !== null &&
+        !stateData.isNextWpNotified && stateData.nextWaypointDistance !== null),
+      tap(([[[currentPosition, routePoints], nextWpIndex, routeDetails], stateData]) => {
+        const currentPositionOnRoute = nearestPointOnLine(lineString(routePoints), point([currentPosition.latitude,
+          currentPosition.longitude
+        ]));
+        const previousPositionOnRoute = nearestPointOnLine(lineString(routePoints), point([stateData.previousPosition.latitude,
+          stateData.previousPosition.longitude
+        ]));
+        const lineBetweenPositions = lineSlice(point([currentPositionOnRoute.geometry.coordinates[0],
+            currentPositionOnRoute.geometry.coordinates[1]
+          ]),
+          point([previousPositionOnRoute.geometry.coordinates[0],
+            previousPositionOnRoute.geometry.coordinates[1]
+          ]),
+          lineString(routePoints));
+        const currentDistance = length(lineBetweenPositions) * 1000;
+        const timeBetweenPositions = (stateData.currentPositionTimeStamp - stateData.previousPositionTimeStamp) / 1000;
+        // console.log(timeBetweenPositions + 'time between positions');
+        // console.log(currentDistance + ' distance between positions');
+        const speed = currentDistance / timeBetweenPositions;
+        const playTime = 4;
+        const preManeuverDistance = 200;
+        // console.log(speed + ' this is the speed');
+        const preNotificationDistance = (speed * playTime) + preManeuverDistance;
+        // console.log(`prenotification distance is ${preNotificationDistance}`);
+        if (stateData.nextWaypointDistance <= preNotificationDistance) {
+          const distanceNotification = Math.min(preManeuverDistance, stateData.nextWaypointDistance);
 
-  indexWaypoints(routeData) {
 
+          // console.log('notification for leg # ' + nextWpIndex);
+          // console.log(`distance to next waypoint: ${stateData.nextWaypointDistance}`);
+          this.store.dispatch(new SetIsNextWpNotifiedAction({ isNextWpNotified: true }));
```

Comment:
For better flow move the dispatch to the effect return statement (and remove dispatch: false).
This means that the effect results in state update.